api/python/t4/formats.py

```diff
@@ -0,0 +1,418 @@
+        handled_types=[bytes],
+).register()
+
+
+Format('json',
+       serializer=lambda obj, **kwargs: json.dumps(obj, **kwargs).encode('utf-8'),
+       deserializer=lambda bytes_obj, **kwargs: json.loads(bytes_obj.decode('utf-8'), **kwargs),
+       handled_extensions=['json'],
+       handled_types=[dict, list, int, float, str, tuple, type(None)]
+).register()
+
+
+Format('unicode',  # utf-8 instead?
+       serializer=lambda s: s.encode('utf-8'),
+       deserializer=lambda b: b.decode('utf-8'),
+       handled_extensions=['txt', 'md', 'rst'],
+       handled_types=[text_type],
+).register()
+
+
+class NumpyFormat(Format):
+    def __init__(self, name, handled_extensions=None, **kwargs):
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+
+        if 'npy' not in handled_extensions:
+            handled_extensions.append('npy')
+        if 'npz' not in handled_extensions:
+            handled_extensions.append('npz')
+
+        super().__init__(name, handled_extensions=handled_extensions, **kwargs)
+
+    def handles_obj(self, obj):
+        # If this is a numpy object, numpy must be loaded.
+        if 'numpy' not in sys.modules:
+            return False
+        import numpy as np
+        if np.ndarray not in self._handled_types:
+            self._handled_types.append(np.ndarray)
+        return super().handles_obj(obj)
+
+    def _deserializer(self, bytes_obj):
+        import numpy as np
+
+        buf = io.BytesIO(bytes_obj)
+        return np.load(buf, allow_pickle=False)
+
+    def _serializer(self, obj):
+        import numpy as np
+        buf = io.BytesIO()
+        np.save(buf, obj, allow_pickle=False)
+        return buf.getvalue()
+
+
+NumpyFormat('numpy').register()
+
+
+class ParquetFormat(Format):
+    def __init__(self, name, handled_extensions=None, **kwargs):
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+
+        if 'parquet' not in handled_extensions:
+            handled_extensions.append('parquet')
+        super().__init__(name, handled_extensions=handled_extensions, **kwargs)
+
+    def handles_obj(self, obj):
+        # don't load pyarrow or pandas unless we actually have to use them..
+        if 'pandas' not in sys.modules:
+            return False
+        import pandas as pd
+        if pd.DataFrame not in self._handled_types:
+            self._handled_types.append(pd.DataFrame)
+        return super().handles_obj(obj)
+
+    def _deserializer(self, bytes_obj):
+        import pyarrow as pa
+        from pyarrow import parquet
+
+        buf = io.BytesIO(bytes_obj)
+        table = parquet.read_table(buf)
+        try:
+            obj = pa.Table.to_pandas(table)
+        except AssertionError:
+            # Try again to convert the table after removing
+            # the possibly buggy Pandas-specific metadata.
+            # XXX: Can we detect this during serialization and store it separately?
```

Comment:
Nope... it might not have been created by us. In any case, Pyarrow raising an assertion error on user input is a pyarrow bug - but they're not in a rush to fix it.

Reply:
Thanks for the reply.